datafusion/common/src/nested_struct.rs

```diff
@@ -159,14 +200,23 @@ pub fn cast_column(source_col: &ArrayRef, target_field: &Field) -> Result<ArrayR
 pub fn validate_struct_compatibility(
     source_fields: &[FieldRef],
     target_fields: &[FieldRef],
-) -> Result<bool> {
+) -> Result<()> {
     // Check compatibility for each target field
     for target_field in target_fields {
         // Look for matching field in source by name
         if let Some(source_field) = source_fields
             .iter()
             .find(|f| f.name() == target_field.name())
         {
+            // Ensure nullability is compatible. It is invalid to cast a nullable
+            // source field to a non-nullable target field as this may discard
+            // null values.
+            if source_field.is_nullable() && !target_field.is_nullable() {
```

Comment:
we might want to control this behaviour with cast options. But it is beyond scope of this PR